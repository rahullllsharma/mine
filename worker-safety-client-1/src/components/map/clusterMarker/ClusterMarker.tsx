import type { RiskLevel } from "@/components/riskBadge/RiskLevel";
import { Circle } from "./components/Circle";
import { calculateFractionsPerRisk } from "./utils/fractions";

type ClusterMarkerPropsKeys = Lowercase<
  Exclude<RiskLevel, RiskLevel.RECALCULATING>
>;

type ClusterMarkerProps = Partial<Record<ClusterMarkerPropsKeys, number>> & {
  size?: number;
};

function ClusterMarker({
  size = 50,
  ...scores
}: ClusterMarkerProps): JSX.Element {
  // total segment size based off the component size.
  const segmentSize = size * 0.5;

  const totalsAsAssoc = Object.entries(scores);
  const total = totalsAsAssoc.reduce(
    (acc, [, value]) => acc + (value > 0 ? value : 0),
    0
  );

  // calculate each arc per score
  const totalsPerFraction = calculateFractionsPerRisk(totalsAsAssoc, total);

  return (
    <div className="inline-block shadow-20 rounded-full max-w-[400px] h-auto">
      <svg
        viewBox={`0 0 ${size} ${size}`}
        width={size}
        height={size}
        xmlns="http://www.w3.org/2000/svg"
      >
        <g type="rect" className="transform rotate-[330deg] origin-center">
          {totalsPerFraction.map(({ risk, ...circleProps }) => (
            <Circle
              key={risk}
              segmentSize={segmentSize}
              segmentTotal={totalsPerFraction.length}
              {...circleProps}
            />
          ))}
        </g>
        <text
          x="50%"
          y="52%"
          dominantBaseline="middle"
          textAnchor="middle"
          className="text-base bg-neutral-shade-100 font-semibold"
        >
          {total < 999 ? total : "+999"}
        </text>
      </svg>
    </div>
  );
}

export default ClusterMarker;
