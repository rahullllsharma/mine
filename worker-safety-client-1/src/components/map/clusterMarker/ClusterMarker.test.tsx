import { render, screen } from "@testing-library/react";
import ClusterMarker from "./ClusterMarker";

describe(ClusterMarker.name, () => {
  it("render correctly", () => {
    const { asFragment } = render(
      <ClusterMarker high={25} medium={25} low={25} unknown={25} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it.each`
    high  | medium       | low          | unknown
    ${25} | ${25}        | ${25}        | ${25}
    ${25} | ${25}        | ${25}        | ${undefined}
    ${25} | ${25}        | ${undefined} | ${undefined}
    ${25} | ${undefined} | ${undefined} | ${undefined}
    ${25} | ${25}        | ${25}        | ${25}
    ${25} | ${25}        | ${25}        | ${0}
    ${25} | ${25}        | ${0}         | ${0}
    ${25} | ${0}         | ${0}         | ${0}
  `("should include a total and the risks passed", props => {
    const params = Object.entries(props) as [string, number | undefined][];
    const total = params.reduce((acc, [, val]) => acc + (val ?? 0), 0);
    const totalRiskUsed = params.reduce((acc, [, v]) => acc + (v ? 1 : 0), 0);

    render(<ClusterMarker {...props} />);

    expect(document.querySelectorAll("path")).toHaveLength(totalRiskUsed);

    // Make sure the proper risk exists in the document
    params
      .filter(([, value]) => !!value)
      .forEach(([param]) => {
        expect(
          document.querySelector(`.text-risk-${param}`)
        ).toBeInTheDocument();
      });

    screen.getByText(total);
  });
});
