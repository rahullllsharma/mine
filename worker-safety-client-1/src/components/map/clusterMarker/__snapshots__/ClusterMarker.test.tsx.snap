// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClusterMarker render correctly 1`] = `
<DocumentFragment>
  <div
    class="inline-block shadow-20 rounded-full max-w-[400px] h-auto"
  >
    <svg
      height="50"
      viewBox="0 0 50 50"
      width="50"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        class="transform rotate-[330deg] origin-center"
        type="rect"
      >
        <path
          class="text-risk-high"
          d="M 41.958588854417016 23.81413994634987 A 17 17 0 0 0 26.18586005365013 8.041411145582988"
          fill="none"
          stroke="currentColor"
          stroke-width="5"
        />
        <path
          class="text-risk-medium"
          d="M 26.18586005365013 41.958588854417016 A 17 17 0 0 0 41.958588854417016 26.18586005365013"
          fill="none"
          stroke="currentColor"
          stroke-width="5"
        />
        <path
          class="text-risk-low"
          d="M 8.041411145582988 26.185860053650135 A 17 17 0 0 0 23.81413994634987 41.958588854417016"
          fill="none"
          stroke="currentColor"
          stroke-width="5"
        />
        <path
          class="text-risk-unknown"
          d="M 23.784543446344323 8.043506690173956 A 17 17 0 0 0 8.041411145582988 23.81413994634988"
          fill="none"
          stroke="currentColor"
          stroke-width="5"
        />
      </g>
      <text
        class="text-base bg-neutral-shade-100 font-semibold"
        dominant-baseline="middle"
        text-anchor="middle"
        x="50%"
        y="52%"
      >
        100
      </text>
    </svg>
  </div>
</DocumentFragment>
`;
